File: DSA/Aanidhay Aggarwal/Day6_q1.java
import java.util.Vector;

class ListNode {
    int val;
    ListNode next;

    ListNode(int val) {
        this.val = val;
        this.next = null;
    }
}

public class CheckCycle {

    public static String checkCycle(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;

            if (slow == fast) {  // A cycle exists
                return "cycle";
            }
        }

        return "no cycle";  // No cycle found
    }

    public static ListNode createLinkedList(Vector<Integer> values) {
        ListNode head = new ListNode(values.get(0));
        ListNode current = head;

        for (int i = 1; i < values.size(); ++i) {
            current.next = new ListNode(values.get(i));
            current = current.next;
        }

        return head;
    }

    public static void main(String[] args) {
        Vector<Integer> values = new Vector<>();
        values.add(1);
        values.add(2);
        values.add(3);
        values.add(4);
        values.add(2);

        ListNode head = createLinkedList(values);

        String result = checkCycle(head);
        System.out.println(result);  // Output: cycle
    }
}

File: DSA/saumya/Day6_Q1.py
class Node:
  def __init__(self, data):
    self.data = data
    self.next = None

def check_cycle(head):

  slow, fast = head, head
  while fast and fast.next:
    slow = slow.next
    fast = fast.next.next
    if slow == fast:
      return "cycle"
  return "no cycle"

# Test cases
# Creating a cycle
head = Node(1)
head.next = Node(2)
head.next.next = Node(3)
head.next.next.next = Node(4)
head.next.next.next.next = head.next

result = check_cycle(head)
print(result)  # Output: cycle

# Creating a non-cycle
head = Node(1)
head.next = Node(2)
head.next.next = Node(3)

result = check_cycle(head)
print(result)  # Output: no cycle
